### ISSUE TITLE
Error Observable cancelled prematurely get's thrown

### ISSUE THREAD

**Issue by capaj** [apollographql/apollo-client#7608]
Sorry that I cannot provide more info.
I may get some time to try and reproduce this in codesandbox in the future, but for now I just wanted to have a linkable bug reference to put into our codebase.
For now we have this hackfix in ErrorBoundary component to get around this:
```ts

  componentDidCatch(error: Error) {
    captureException(error)

    if (error.message === 'Observable cancelled prematurely') {
      console.warn(error) // this is an annoying error getting thrown by apollo client 3.3.x. Some future versions might be okay, but it's the safest to leave this here
      // easiest to get this is to use login link on a user profile page and open it in incognito window, There it happens 100% of the times
      location.reload() // slightly hacky, but we can't think of a better hotfix ATM
      return
    }
    this.setState({ error })
  }
  ```

**Intended outcome:**
no error is thrown no matter how we mount/dismount providers.
If this indeed happens when a provider is unmounted by react then it should be caught by apollo-client because at that point we really don't care about these observables and their values at all.

**Actual outcome:**
this error is thrown:
```
Uncaught Error: Observable cancelled prematurely
  | ApolloError | @ | index.js:26
  | (anonymous) | @ | QueryManager.js:535
  | (anonymous) | @ | asyncMap.js:11
  | (anonymous) | @ | asyncMap.js:11
  | notifySubscription | @ | Observable.js:140
  | flushSubscription | @ | Observable.js:121
  | (anonymous) | @ | Observable.js:174
  | (anonymous) | @ | Observable.js:73
  | Promise.then (async) |   |  
  | onNewData | @ | useBaseQuery.js:18
  | error | @ | QueryData.js:240
  | notifySubscription | @ | Observable.js:140
  | onNotify | @ | Observable.js:179
  | error | @ | Observable.js:240
  | (anonymous) | @ | iteration.js:4
  | iterateObserversSafely | @ | iteration.js:4
  | error | @ | ObservableQuery.js:26
  | (anonymous) | @ | iteration.js:4
  | iterateObserversSafely | @ | iteration.js:4
  | error | @ | Concast.js:35
  | notifySubscription | @ | Observable.js:140
  | onNotify | @ | Observable.js:179
  | error | @ | Observable.js:240
  | (anonymous) | @ | asyncMap.js:19
  | Promise.then (async) |   |  
  | (anonymous) | @ | asyncMap.js:11
  | notifySubscription | @ | Observable.js:140
  | flushSubscription | @ | Observable.js:121
  | (anonymous) | @ | Observable.js:174
  | (anonymous) | @ | Observable.js:73
  | Promise.then (async) |   |  
  | enqueue | @ | Observable.js:71
  | onNotify | @ | Observable.js:173
  | error | @ | Observable.js:240
  | ../node_modules/@apollo/client/utilities/observables/Concast.js.Concast.deliverLastMessage | @ | Concast.js:86
  | ../node_modules/@apollo/client/utilities/observables/Concast.js.Concast.addObserver | @ | Concast.js:97
  | (anonymous) | @ | Concast.js:11
  | Subscription | @ | Observable.js:197
  | subscribe | @ | Observable.js:279
  | (anonymous) | @ | asyncMap.js:37
  | Subscription | @ | Observable.js:197
  | subscribe | @ | Observable.js:279
  | complete | @ | Concast.js:56
  | notifySubscription | @ | Observable.js:145
  | onNotify | @ | Observable.js:179
  | complete | @ | Observable.js:245
  | (anonymous) | @ | Observable.js:593
  | (anonymous) | @ | Observable.js:73
  | Promise.then (async) |   |  
  | enqueue | @ | Observable.js:71
  | (anonymous) | @ | Observable.js:585
  | Subscription | @ | Observable.js:197
  | subscribe | @ | Observable.js:279
  | complete | @ | Concast.js:56
  | ../node_modules/@apollo/client/utilities/observables/Concast.js.Concast.start | @ | Concast.js:79
  | Concast | @ | Concast.js:71
  | ../node_modules/@apollo/client/core/QueryManager.js.QueryManager.fetchQueryObservable | @ | QueryManager.js:580
  | (anonymous) | @ | ObservableQuery.js:312
  | ../node_modules/@apollo/client/core/Reobserver.js.Reobserver.reobserve | @ | Reobserver.js:18
  | ../node_modules/@apollo/client/core/ObservableQuery.js.ObservableQuery.reobserve | @ | ObservableQuery.js:317
  | ../node_modules/@apollo/client/core/ObservableQuery.js.ObservableQuery.onSubscribe | @ | ObservableQuery.js:294
  | (anonymous) | @ | ObservableQuery.js:13
  | Subscription | @ | Observable.js:197
  | subscribe | @ | Observable.js:279
  | ../node_modules/@apollo/client/react/data/QueryData.js.QueryData.startQuerySubscription | @ | QueryData.js:220
  | ../node_modules/@apollo/client/react/data/QueryData.js.QueryData.getExecuteResult | @ | QueryData.js:150
  | ../node_modules/@apollo/client/react/data/QueryData.js.QueryData.execute | @ | QueryData.js:98
  | (anonymous) | @ | useBaseQuery.js:35
  | useDeepMemo | @ | useDeepMemo.js:6
  | useBaseQuery | @ | useBaseQuery.js:35
  | useQuery | @ | useQuery.js:3
  | useThrowingQuery | @ | useThrowingQuery.ts:18
  | useLoginPageQuery | @ | LoginPage.codegen.tsx:157
  | OauthLoginButtons | @ | OauthLoginButtons.tsx:36
  | renderWithHooks | @ | react-dom.development.js:14985
  | mountIndeterminateComponent | @ | react-dom.development.js:17811
  | beginWork | @ | react-dom.development.js:19049
  | beginWork$1 | @ | react-dom.development.js:23940
  | performUnitOfWork | @ | react-dom.development.js:22776
  | workLoopSync | @ | react-dom.development.js:22707
  | renderRootSync | @ | react-dom.development.js:22670
  | performSyncWorkOnRoot | @ | react-dom.development.js:22293
  | (anonymous) | @ | react-dom.development.js:11327
  | unstable_runWithPriority | @ | scheduler.development.js:646
  | runWithPriority$1 | @ | react-dom.development.js:11276
  | flushSyncCallbackQueueImpl | @ | react-dom.development.js:11322
  | flushSyncCallbackQueue | @ | react-dom.development.js:11309
  | scheduleUpdateOnFiber | @ | react-dom.development.js:21893
  | dispatchAction | @ | react-dom.development.js:16139
  | Promise.then (async) |   |  
  | onNewData | @ | useBaseQuery.js:18
  | next | @ | QueryData.js:230
  | notifySubscription | @ | Observable.js:135
  | onNotify | @ | Observable.js:179
  | next | @ | Observable.js:235
  | (anonymous) | @ | iteration.js:4
  | iterateObserversSafely | @ | iteration.js:4
  | next | @ | ObservableQuery.js:21
  | (anonymous) | @ | iteration.js:4
  | iterateObserversSafely | @ | iteration.js:4
  | next | @ | Concast.js:24
  | notifySubscription | @ | Observable.js:135
  | onNotify | @ | Observable.js:179
  | next | @ | Observable.js:235
  | (anonymous) | @ | Observable.js:589
  | (anonymous) | @ | Observable.js:73
  | Promise.then (async) |   |  
  | enqueue | @ | Observable.js:71
  | (anonymous) | @ | Observable.js:585
  | Subscription | @ | Observable.js:197
  | subscribe | @ | Observable.js:279
  | complete | @ | Concast.js:56
  | ../node_modules/@apollo/client/utilities/observables/Concast.js.Concast.start | @ | Concast.js:79
  | Concast | @ | Concast.js:71
  | ../node_modules/@apollo/client/core/QueryManager.js.QueryManager.fetchQueryObservable | @ | QueryManager.js:580
  | (anonymous) | @ | ObservableQuery.js:312
  | ../node_modules/@apollo/client/core/Reobserver.js.Reobserver.reobserve | @ | Reobserver.js:18
  | ../node_modules/@apollo/client/core/ObservableQuery.js.ObservableQuery.reobserve | @ | ObservableQuery.js:317
  | listeners.add.oqListener | @ | QueryInfo.js:103
  | (anonymous) | @ | QueryInfo.js:115
  | ../node_modules/@apollo/client/core/QueryInfo.js.QueryInfo.notify | @ | QueryInfo.js:115
  | (anonymous) | @ | QueryManager.js:447
  | ../node_modules/@apollo/client/core/QueryManager.js.QueryManager.broadcastQueries | @ | QueryManager.js:447
  | ../node_modules/@apollo/client/core/QueryManager.js.QueryManager.stopQuery | @ | QueryManager.js:433
  | (anonymous) | @ | QueryManager.js:332
  | Promise.finally (async) |   |  
  | ../node_modules/@apollo/client/core/QueryManager.js.QueryManager.query | @ | QueryManager.js:332
  | ../node_modules/@apollo/client/core/ApolloClient.js.ApolloClient.query | @ | ApolloClient.js:133
  | (anonymous) | @ | loadCustomFonts.ts:113
  | (anonymous) | @ | loadCustomFonts.codegen.tsx:64
  | __async | @ | loadCustomFonts.codegen.tsx:64
  | loadCustomFonts | @ | loadCustomFonts.ts:112
  | transitionAppStateToLoggedIn | @ | sessionStore.ts:289
  | AuthSuccess | @ | AuthSuccess.tsx:34
  | renderWithHooks | @ | react-dom.development.js:14985
  | updateFunctionComponent | @ | react-dom.development.js:17356
  | beginWork | @ | react-dom.development.js:19063
  | beginWork$1 | @ | react-dom.development.js:23940
  | performUnitOfWork | @ | react-dom.development.js:22776
  | workLoopSync | @ | react-dom.development.js:22707
  | renderRootSync | @ | react-dom.development.js:22670
  | performSyncWorkOnRoot | @ | react-dom.development.js:22293
  | (anonymous) | @ | react-dom.development.js:11327
  | unstable_runWithPriority | @ | scheduler.development.js:646
  | runWithPriority$1 | @ | react-dom.development.js:11276
  | flushSyncCallbackQueueImpl | @ | react-dom.development.js:11322
  | flushSyncCallbackQueue | @ | react-dom.development.js:11309
  | scheduleUpdateOnFiber | @ | react-dom.development.js:21893
  | dispatchAction | @ | react-dom.development.js:16139
  | Promise.then (async) |   |  
  | onNewData | @ | useBaseQuery.js:18
  | next | @ | QueryData.js:230
  | notifySubscription | @ | Observable.js:135
  | onNotify | @ | Observable.js:179
  | next | @ | Observable.js:235
  | (anonymous) | @ | iteration.js:4
  | iterateObserversSafely | @ | iteration.js:4
  | next | @ | ObservableQuery.js:21
  | (anonymous) | @ | iteration.js:4
  | iterateObserversSafely | @ | iteration.js:4
  | next | @ | Concast.js:24
  | notifySubscription | @ | Observable.js:135
  | onNotify | @ | Observable.js:179
  | next | @ | Observable.js:235
  | (anonymous) | @ | asyncMap.js:13
  | Promise.then (async) |   |  
  | (anonymous) | @ | asyncMap.js:11
  | notifySubscription | @ | Observable.js:135
  | onNotify | @ | Observable.js:179
  | next | @ | Observable.js:235
  | (anonymous) | @ | iteration.js:4
  | iterateObserversSafely | @ | iteration.js:4
  | next | @ | Concast.js:24
  | notifySubscription | @ | Observable.js:135
  | onNotify | @ | Observable.js:179
  | next | @ | Observable.js:235
  | next | @ | bundle.esm.js:29
  | notifySubscription | @ | Observable.js:135
  | onNotify | @ | Observable.js:179
  | next | @ | Observable.js:235
  | (anonymous) | @ | batching.js:84
  | (anonymous) | @ | batching.js:84
  | next | @ | batching.js:82
  | notifySubscription | @ | Observable.js:135
  | onNotify | @ | Observable.js:179
  | next | @ | Observable.js:235
  | (anonymous) | @ | batchHttpLink.js:75
  | Promise.then (async) |   |  
  | (anonymous) | @ | batchHttpLink.js:74
  | Subscription | @ | Observable.js:197
  | subscribe | @ | Observable.js:279
  | ../node_modules/@apollo/client/link/batch/batching.js.OperationBatcher.consumeQueue | @ | batching.js:72
  | (anonymous) | @ | batching.js:105


```
**How to reproduce the issue:**
I actually don't know.
This error is thrown on startup of our closed source app, so I cannot share the code.
The app actually has 2 graphQL endpoints and we switch between them depending on the route where the user is and I think this error get's  thrown when one apollo provider is unmounted, but I cannot be sure. 

 
**Versions**
```
  System:
    OS: Linux 5.4 Ubuntu 20.04.1 LTS (Focal Fossa)
  Binaries:
    Node: 12.18.4 - ~/.nvm/versions/node/v12.18.4/bin/node
    Yarn: 1.22.4 - ~/.yarn/bin/yarn
    npm: 6.14.6 - ~/.nvm/versions/node/v12.18.4/bin/npm
  Browsers:
    Chrome: 87.0.4280.141
    Firefox: 84.0.2
  npmPackages:
    @apollo/client: ^3.3.7 => 3.3.7 
    @apollo/link-batch-http: ^2.0.0-beta.3 => 2.0.0-beta.3 
    @apollo/link-context: ^2.0.0-beta.3 => 2.0.0-beta.3 
    @apollo/link-error: ^2.0.0-beta.3 => 2.0.0-beta.3 
    @apollo/link-schema: ^2.0.0-beta.3 => 2.0.0-beta.3 
    @apollo/react-components: ^3.1.5 => 3.1.5 
    @apollo/react-testing: ^3.1.4 => 3.1.4 
    apollo-client: ^2.6.10 => 2.6.10 
    apollo-server: ^2.19.2 => 2.19.2 
    apollo-server-express: ^2.19.2 => 2.19.2 
    react-apollo-network-status: ^5.0.1 => 5.0.1 

```

**Comment by JKulhavy91** [comment]
I am experiencing same issue. This had started with version Apollo Client 3.3.0. For now I had downgraded to Apollo Client 3.2.9.  .  I think that this issue is connected with this feature "Unsubscribing the last observer from an ObservableQuery will once again unsubscribe from the underlying network Observable in all cases, as in Apollo Client 2.x, allowing network requests to be cancelled by unsubscribing." Witch was implemented in  Apollo Client 3.3.0

**Comment by jochienabuurs** [comment]
Having the same issue here. With me, it also seems to happen when the provider gets unmountend and mounted again.

**Comment by PowerKiKi** [comment]
Experiencing the same thing in an Angular project, when navigating away from a page where a query is happening.

**Comment by kelvin2200** [comment]
We're having the same issue on v.3.3.9. Reverting to 3.2.9 is not an option, because it starts throwing Invariant Violations. Any insight on how to fix this? I can confirm that the Apollo Client Provider is being mounted only ONCE and never re-mounted.

**Comment by davidgilbertson** [comment]
Thanks @capaj for posting your workaround, very useful.

To the Apollo folks, it seems to me like throwing an error and breaking the user's app is a bit extreme for this type of issue (especially since it appear to only occur in production mode).

I have no knowledge of this codebase, but could [this line](https://github.com/apollographql/apollo-client/blob/main/src/utilities/observables/Concast.ts#L62) not pass `true` as a second parameter to prevent throwing errors? Or are errors being thrown as a method of communication, and accidentally being allowed out into userland?

**Comment by Jscripter-pk** [comment]
We are also facing this issue in our React app
**Explanation** :
We are using query to fetch our user .... while querying user we get this error sometimes

**Comment by alperen** [comment]
I was having this problem, I read and research several things about that issue. Unfortunately, the issue is specific to your approach and codebase. _You may have a memory leak or unexpected behavior on your product as a code._ Already it proofs that because all these comments are pointing to different complaints.

**Comment by patricknazar** [comment]
We are using apollo-angular. We use a `switchMap` to start a new query when the variables change. We watch the query with `.valueChanges`. This used to be ok but now it seems it doesn't like that we've done that. It cancels the new query, and errors with "Observable cancelled prematurely". I suspect that other's are using router params to do the same thing and are getting the same error. Using `mergeMap` instead stops the error, but naturally causes new problems, such as wasting network bandwidth and returning values for old queries we don't want anymore.

I've worked around the issue by using a `takeUntil` which will unsub the switched inner observable which a condition arises (in our case an ngrx action), but I don't like it.

I'd like to know how others are dealing with the "watching a query that has changing variables" use case.

**Comment by brainkim** [comment]
This definitely looks like a problem on our end and I apologize for all the grief it’s caused. I currently have a lot on my plate, but I will make sure to look at this soon. In the meantime, if anyone created a smallish reproduction using our error template (https://github.com/apollographql/react-apollo-error-template), that would probably magically move this issue up in my TODOs 😇

**Comment by patricknazar** [comment]
It seems I only get this error if `returnPartialData = true`. Can anyone else confirm this in theirs? I'm trying to put together a repro.

**Comment by bchTrackunit** [comment]
This started for me when i used a hook together with a LazyQuery that refetches in a useEffect.

```javascript

const { someIds } = useSomeHook();

const [getItems, { loading, data }] = useItemsLazyQuery();

React.useEffect(() => {
    getItems({
      variables: {
        ids: someIds,
      },
    });
  }, [someIds]);

```

if i remove the code from the hook, and move the logic next to the query, it will work fine!

**Comment by 3nvi** [comment]
For us, it happens when you unmount and remount the sasme component really fast.

**Comment by patricknazar** [comment]
Any progress? This issue is a huge problem for me.

**Comment by lilred** [comment]
We're seeing this issue in production as well.

**Comment by cornedor** [comment]
Switching the `Promise.resolve` for a `setTimeout(()=>{}, 0)` at https://github.com/apollographql/apollo-client/blob/main/src/utilities/observables/Concast.ts#L181 resolves this issue for me. However, I'm not sure if this is a real fix, or that the error is simple suppressed now.

**Comment by razor-x** [comment]
I am seeing this when using react-router `<Redirect to={someRoute} />`. In my case, I run a GQL query to check the status of the user's account, and in some cases I will redirect the user to a different page. The redirect will happen (url bar updated), but then the error will be thrown at some point during the rendering of the new route.

Using [capaj's workaround](https://github.com/apollographql/apollo-client/issues/7608#issue-791969580) fixes the issue for now, but it's not an idea UX.

**Comment by razor-x** [comment]
If it helps, I started getting this error after updating `@apollo/client 3.3.21 => 3.4.0` and still get it on the latest version 3.4.5.

**Comment by nikolnikov** [comment]
I noticed this starting to happen in my Angular app. It generally happens on sign in. You immediately get booted out of the app and it throws the "Observable cancelled prematurely" message. Unfortunately, we cannot reproduce with certainty so it's been a nightmare to try and replicate, much less fix.

**Comment by jdehaan** [comment]
We are seen this as well and happens typically during a storm of queries for the same data with cache-first (should resolve to cache). Maybe that helps getting an idea for some repro. Here same thing sometimes works, sometimes not :-/

**Comment by benjamn** [comment]
@cornedor (and @matthewrc) I actually think the `setTimeout` trick is good/appropriate, since it takes care of the "Unhandled Promise Rejection" version of this error, which clearly should not be fatal (and there's already a small delay before unsubscribing anyway).

Partial(?) fix incoming: #8676

**Comment by hwillson** [comment]
Let us know if things are fixed with `@apollo/client@latest` - thanks!

**Comment by nikolnikov** [comment]
v3.4.10 still boots me out of my app and throws the observable error. I downgraded to the latest build, which seems to fix the issue, build 3.2.9.

**Comment by psrcek** [comment]
This happened for me when I tried to resubscribe to a subscription I've already unsubscribed from. You need to run `client.subscribe({...})` every time you want to make a subscription to it, it doesn't work like a regular observable where you can subscribe and unsubscribe to it at will (but IMO it should!)

**Comment by benjamn** [comment]
Unfortunately, I'm out of ideas for easy fixes, unless someone can provide a [reproduction](https://github.com/apollographql/react-apollo-error-template/). 🙏

**Comment by Abdallah-khalil** [comment]
in our angular , hasura app we use  [graphql-ws  and apollo](https://github.com/enisdenjo/graphql-ws#apollo-client)  we have 2 real time subscriptions that happens on app intialization , with some other queries running also on some component initializations and i was getting this error, but i did a small hack and wrapped those subscriptions into a setTimeout function with 1s and i don't see this error any more

**Comment by tienle** [comment]
I just got this same issue with v3.4.10. It seems to come from a subscription from another query's `subscribeToMore` that placed in the useEffect hook. Still trying to investigate the root cause and finding out this thread.

**Comment by felixmeziere** [comment]
Got the error as well....

**Comment by Abdallah-khalil** [comment]
> in our angular , hasura app we use [graphql-ws and apollo](https://github.com/enisdenjo/graphql-ws#apollo-client) we have 2 real time subscriptions that happens on app intialization , with some other queries running also on some component initializations and i was getting this error, but i did a small hack and wrapped those subscriptions into a setTimeout function with 1s and i don't see this error any more

Started to get this randomly again :no_good:

**Comment by hwillson** [comment]
As mentioned in https://github.com/apollographql/apollo-client/issues/7608#issuecomment-916442136 if anyone can provide a small runnable reproduction that demonstrates this, that would be great. Thanks!

**Comment by mashtii85** [comment]
In react I could manage to solve it, exchange useQuery with useLazyQuery inside useEffect

**Comment by andreialecu** [comment]
Ran into this recently, and I'm on `apollo-client` 3.4.17. I'm able to reproduce it fairly consistently, here's the stack trace:

```
Error: Observable cancelled prematurely
  at removeObserver(webpack:/./node_modules/@apollo/client/utilities/observables/Concast.js:110:33)
  at cleanup(webpack:/./node_modules/@apollo/client/utilities/observables/Concast.js:13:47)
  at cleanupSubscription(webpack:/./node_modules/zen-observable-ts/module.js:88:7)
  at call(webpack:/./node_modules/zen-observable-ts/module.js:203:7)
  at cleanupSubscription(webpack:/./node_modules/zen-observable-ts/module.js:93:21)
  at unsubscribe(webpack:/./node_modules/zen-observable-ts/module.js:203:7)
  at unsubscribe(webpack:/./node_modules/rxjs/_esm2015/internal/Subscription.js:51:29)
  at unsubscribe(webpack:/./node_modules/rxjs/_esm2015/internal/Subscription.js:51:29)
  at unsubscribe(webpack:/./node_modules/rxjs/_esm2015/internal/Subscriber.js:69:15)
  at unsubscribe(webpack:/./node_modules/rxjs/_esm2015/internal/Subscription.js:51:29)
  at unsubscribe(webpack:/./node_modules/rxjs/_esm2015/internal/Subscriber.js:69:15)
  at _innerSub(webpack:/./node_modules/rxjs/_esm2015/internal/operators/switchMap.js:39:31)
  at _next(webpack:/./node_modules/rxjs/_esm2015/internal/operators/switchMap.js:34:14)
  at next(webpack:/./node_modules/rxjs/_esm2015/internal/Subscriber.js:49:18)
  at next(webpack:/./node_modules/rxjs/_esm2015/internal/Subject.js:39:25)
  at next(webpack:/./node_modules/rxjs/_esm2015/internal/ReplaySubject.js:31:15)
  at setCurrentClub(webpack:/./src/app/admin/admin-shared/services/core.service.ts:201:42)
  at apply(webpack:/./src/app/admin/admin.component.ts:150:24)
```

Relevant code from `admin.component.ts` (simplified for brevity):

```ts
  private reloadClubs(user: AdminUser) {
    return this.getClubsGQL.then((clubs) => {
      this.coreService.setCurrentClub(
        clubs[0],
      )
    }
  }
```

The error occurs on this line in `setCurrentClub`:

```ts
  public setCurrentClub(club: Club) {
    this.currentClub = club;
    this.currentClub$.next(club);
    this.globalStateService.currentClub$.next(club); // error here
  }
```

GlobalStateService is simply:

```ts
@Injectable({ providedIn: 'root' })
export class GlobalStateService {
  public currentClub$ = new ReplaySubject<Club>(1);
}
```

Then there's a consumer of this in a separate service:

```ts
    this.globalStateService.currentClub$
      .pipe(
        switchMap(() => {
          return this.playerChangedGQL.subscribe();
        }),
        tap((p) => {
          ...
        }),
      )
      .subscribe();
```

Notice how the stack trace mentions this `switchMap` being called.

Also the fix in [#](https://github.com/apollographql/apollo-client/pull/8676) by @benjamn is not part of this stack trace, so that timeout had no chance to be applied.

**Comment by andreialecu** [comment]
The repro in https://github.com/apollographql/apollo-client/issues/7608#issuecomment-830368493 shows a similar issue but I'm getting this without using `returnPartialData`.

Instead, `playerChangedGQL` in the last step is a GQL Subscription, which may be relevant.

**Comment by andreialecu** [comment]
> Unfortunately, I'm out of ideas for easy fixes, unless someone can provide a [reproduction](https://github.com/apollographql/react-apollo-error-template/). 🙏

Repro for the issue in my previous post:

https://stackblitz.com/edit/apollo-bug-test-xf2dxs?file=src/app/exchange-rates/exchange-rates.service.ts

<img width="618" alt="Screenshot 2021-11-16 at 12 38 32" src="https://user-images.githubusercontent.com/697707/141970188-581363fd-0054-40ca-a18e-c57f640ba77f.png">

Note that it seems to happen just when `ReplaySubject` is involved. If I replace it with `BehaviorSubject` it seems to work fine. 

Relevant code is in `exchange-service.ts`.

**Comment by MassivDash** [comment]
I have came across this Error with the newest client apollo client while using ```useSubscription```, this happens when you unmount and mount a component "fast". In my case I had a pageWrapper using the subscription, it was present in  multiple  "pages" react route components, and showed up while navigating through the app.  Lifting the pageWrapper above the <switch> solved the issue for me.

**Comment by DavidPlonsky** [comment]
Hi @ all. Came across this error on using `setVariables(myParams)`. As i used `refetch(myParams)` it solved the problem. Maybe this helps some of you.

**Comment by youhaowei** [comment]
Ran into this where the component with useQuery was dismounted and remounted when the route changed, the error is thrown with fetch policy "cache-and-network", switching to "cache-first" and the error would go away for me.

**Comment by meetzaveri** [comment]
I was facing this issue using `useSubsrciption` hook. Was using `useSubscription` hook into a component which was mounting and then unmounting.

Figured out the unmounting reason. So when I fixed and stopped unmounting behavior, the problem was solved for me.

For those who want to check mount and unmount behavior for your component, I am happy to provide you example snippet
```js
 // Checking mounting and unmounting if it happens in your component
 useEffect(() => {
            console.log(
                'on mount Element wrapper',
                new Date().getMinutes(), // logs minutes
                new Date().getSeconds() // logs seconds so that you can check diff in browser
            )
            return () => {
                console.log(
                    'on unmount Element wrapper',
                    new Date().getMinutes(), // logs minutes
                    new Date().getSeconds() // logs seconds so that you can check diff in browser
                )
            }
        }, [])
```

**Comment by MaxiSantos** [comment]
I'm having the same issue `Error Observable cancelled prematurely get's thrown` It looks its related with what major people say about unmounting and mounting very fast. Can't change that behavior right now.. I'd like to have an option to update my subscription so I can avoid having to unmount and create a new one. But for now I'm delaying the mount of the component that handle my subscription with a Delay wrapper.

https://stackoverflow.com/a/51771466/3120585

**Comment by joeyorlando** [comment]
not to repeat what others have said above but downgrading from `3.5.8` to `3.2.9` made this magically disappear for me (was also getting the `Error Observable cancelled prematurely get's thrown` error when using `useSubscription`).

In my case I'm using a `WebsocketLink`. The connection is fine, I can see data being sent/received by my ws connection and the connection remains open/sends keep-alive signals. Something internal within the `useSubscription` hook seemed to be unhappy and throw the error:

```tsx
// I'm using graphql-codegen + the typescript-react-apollo plugin to generate react hooks for me
import { useCurrentUserUpdatedSubscription } from './currentUserUpdatedSubscription.generated';

const { Content } = Layout;

const DashboardLayout: FC = ({ children }) => {
  const { data, loading, error } = useCurrentUserUpdatedSubscription({
    onSubscriptionData({ subscriptionData, client: { cache } }) {
      const userUpdates = subscriptionData.data?.currentUserUpdated?.user;

      console.log('data received', subscriptionData);
    },
  });

  console.log(data, loading, error); // 'error' would contain the infamous 'Observable cancelled prematurely get's thrown'

  return (
    <div className="flex flex-col w-full h-full">
      <header>
        <Navbar />
      </header>
      <div className="flex flex-1">
        <Sidebar />
        <Content style={{ padding: '0 50px' }}>
          {/* <Breadcrumbs /> */}
          <AccountSetupMessage />
          <div className="site-layout-content">{children}</div>
        </Content>
      </div>
    </div>
  );
};
```

**Comment by PowerKiKi** [comment]
This issue could get fixed in the refactoring happening in #9331. But we'll need a sign from core team for that PR to progress further.

**Comment by manuFL** [comment]
eagerly waiting for a fix too...

**Comment by lor95** [comment]
waiting for a fix, unfortunately we can't downgrade to `3.2.9`...

**Comment by vincekinyops** [comment]
still experiencing this on version 3.5.7
any workaround fix would be fine

**Comment by adamduren** [comment]
In my case it appeared to be because I was using `subscribeMore` from `useQuery` in an `useEffect`. I called the cancel function returned by `subscribeMore` in the cleanup of `useEffect`. My guess is that there's a race condition and the internal cleanup function runs and conflicts with the external cleanup function.

**Comment by gthau** [comment]
I have noticed in my codebase that it happens when I use Graphql subscriptions directly using Apollo `.subscribe({ query: myGqlSubscription })` (this happens in very few places where I can't use a `watchQuery` with `subscribeToMore`)

The scenario is something like:
```ts
const selectedPage$: Observable<string> = ...;
const getPageData = (pageId) => apolloClient.query({ query: myGqlQuery, variables: { pageId } });
const getUpdatesForPage = (pageId) => apolloClient.subscribe({ query: myGqlSubscription, variables: { pageId } });

const data$ = selectedPage$.pipe(
  switchMap((pageId) => getPageData(pageId)),
);
const updates$ = selectedPage$.pipe(
  switchMap((pageId) => getUpdatesForPage(pageId)),
);


// somewhere else 
data$.subscribe((data) => {
  // do something with data
});
updates$.subscribe((updates) => {
 // do something with updates
});
```
In this scenario, when `selectedPage$` emits its second value, `switchMap` auto-unsubscribes the subscriber and restarts a new subscription. This is not a problem for the Apollo query. However it is a problem for the Apollo subscription. Apparently this happens too fast so the ack from the gql server is received when the subscription is already closed, hence the "Error Observable cancelled prematurely".

The workaround I have found is to:
- explicitely terminating the subscription by using `takeUntil` and a shareReplay source observable
- delay a bit the switchMap so that the subscription can be terminated properly
```ts
// shareReplay the source observable
const selectedPage$: Observable<string> = mySourceObs.pipe(
  shareReplay({ bufferSize: 1, refCount: true }),
);
const getPageData = (page: number) => apolloClient.query({ query: myGqlQuery, variables: { page } });
const getUpdatesForPage = (page: number) => apolloClient.subscribe({ query: myGqlSubscription, variables: { page } }).pipe(
  // by using takeUntil we force the subscription to terminate, combine with the delay(200) on the switchMap, it gives time to properly handle this unsubscription
  // skip(1) is necessary because the observable is replayed so it will always emit its first value immediately, by skipping 1 value we unsubscribe on the page change
  takeUntil(selectedPage$.pipe(skip(1))),
);

const data$ = selectedPage$.pipe(
  // no delay required
  switchMap((pageId) => getPageData(pageId)),
);
const updates$ = selectedPage$.pipe(
  // here we delay to let some time for the unsubscription to be handled properly 
  delay(200),
  switchMap((pageId) => getUpdatesForPage(pageId)),
);

// somewhere else 
data$.subscribe((data) => {
  // do something with data
});
updates$.subscribe((updates) => {
 // do something with updates
});
```

This feels hackish but it does work for me. Hope it helps.

**Comment by mikejackowski** [comment]
Any word on getting this fixed?

**Comment by Jaikant** [comment]
I am seeing this issue while using subscription with react-table.

**Comment by Dashue** [comment]
```  js
  const focused = useIsFocused();
  const {data, loading, error} = useSubscription(
    statusUpdatedSubscription,
    {
      variables: {id: '123'},
      shouldResubscribe: true,
      // skip: !focused,
    },
  );
```

Appears my case was caused by conditionally skipping using the skip parameter.
Commenting this out seems to have made it better.

**Comment by tomaswitek** [comment]
@[hwillson](https://github.com/hwillson) still interested in a reproduction repo? I am getting the same error as all users above when upgrading from `3.4` to `3.5`.

**Comment by mikejackowski** [comment]
As a workaround, for me, I do query -> subscribeToMore. Seems to be getting the job done *for now* but I feel it's hacky and it shouldn't be this way.

**Comment by tomaswitek** [comment]
@mikejackowski it would be a pity to stop using `useSubscription`

I've done some debugging. In the `3.4` version it also rejects the same error, but it somehow isn't propagated into the error return hook variable. So maybe it was wrong in the older versions :)

<img width="1359" alt="Snímek obrazovky 2022-04-27 v 11 40 37" src="https://user-images.githubusercontent.com/340649/165491693-c40f30c1-aedd-430f-9b7e-070956fbc046.png">

**Comment by chegger** [comment]
Did some digging and found something quite interesting:

We recently upgraded to use `react 18.0.0`, using `@apollo/client 3.6.1`.
One interesting thing with react 18 is its [new development-only check in `StrictMode`](https://reactjs.org/blog/2022/03/29/react-v18.html#new-strict-mode-behaviors). It

> will automatically unmount and remount every component, whenever a component mounts for the first time, restoring the previous state on the second mount.

Reading the whole linked paragraph, it also says that this mount/unmount behavior will be a performance optimization feature and that it 

> requires components to be resilient to effects being mounted and destroyed multiple times

This means that this will be a future problem in non-development react apps as well.

Since we use `StrictMode` for our app, this new mount/unmount behavior causes the problem for us for our `useSubscription` hooks. When I remove `StrictMode`, the subscription worked like a charm again! However, aside from debugging, that's not an option because a) we want to use `StrictMode` and b) as described above, this behavior is becoming a feature in react 18.

**Comment by codingwaysarg** [comment]
In our case the issue was that we had two components rendering at the same time and starting the same subscription. While that was happening, one of them used to unmount due to history.replace in a matching condition. That was causing "observable cancelled prematurely".

**Comment by doron-cohen** [comment]
Adding to what people wrote here. We hit this issue with regular queries (not a subscription) and only when developing locally (with `vite`).
The HTTP request was sent and returned but the `data` didn't update. Just as if the request dissapeared. Looking at the `fetch` promise I saw this `AbortSignal`:

```
Error: signal is aborted without reason
    at http://localhost:3001/node_modules/.vite/chunk-PBNBAGNW.js?v=ebd50497:432:26
    at cleanupSubscription (http://localhost:3001/node_modules/.vite/chunk-QRI54YM7.js?v=ebd50497:7560:7)
    at notifySubscription (http://localhost:3001/node_modules/.vite/chunk-QRI54YM7.js?v=ebd50497:7616:5)
    at onNotify (http://localhost:3001/node_modules/.vite/chunk-QRI54YM7.js?v=ebd50497:7641:3)
    at SubscriptionObserver2.complete (http://localhost:3001/node_modules/.vite/chunk-QRI54YM7.js?v=ebd50497:7703:9)
    at http://localhost:3001/node_modules/.vite/chunk-PBNBAGNW.js?v=ebd50497:420:22
```

Downgrading to 3.2.9 like the rest solved the issue.

**Comment by chegger** [comment]
Is there any update to this? This behavior with regards to quick mounts/unmounts seems to be breaking with React 18.

**Comment by iserdmi** [comment]
I receive this error in usual query when I set {fetchPolicy: 'cache-and-network'}, when I replace it to {fetchPolicy: 'network'} error disappeared

**Comment by benjamn** [comment]
I'm cautiously optimistic this problem may be solved once and for all by #9701, which you can test by running
```sh
npm i @apollo/client@beta
```
to get version `3.7.0-alpha.3`. If all goes well, these changes should be safe to release in 3.6.4, not waiting from v3.7, so please let us know if/when you have a chance to test the alpha!

**Comment by ShantanuPhatke** [comment]
> I'm cautiously optimistic this problem may be solved once and for all by #9701, which you can test by running
> 
> ```shell
> npm i @apollo/client@beta
> ```
> 
> to get version `3.7.0-alpha.3`. If all goes well, these changes should be safe to release in 3.6.4, not waiting from v3.7, so please let us know if/when you have a chance to test the alpha!

`3.7.0-alpha.3` resolved the issue for me! Was stuck on v3.2.9 only because the later versions had this issue. Thanks for the fix, hope it is merged in the stable release soon.

**Comment by Nedomas** [comment]
Can also confirm that `3.7.0-alpha.3` fixed the `Observable cancelled prematurely` issue for me too.

**Comment by doflo-dfa** [comment]
> I receive this error in usual query when I set {fetchPolicy: 'cache-and-network'}, when I replace it to {fetchPolicy: 'network'} error disappeared

This is like due to the additional time it takes to do the network side of the fetch

**Comment by ShantanuPhatke** [comment]
> > I'm cautiously optimistic this problem may be solved once and for all by #9701, which you can test by running
> > ```shell
> > npm i @apollo/client@beta
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to get version `3.7.0-alpha.3`. If all goes well, these changes should be safe to release in 3.6.4, not waiting from v3.7, so please let us know if/when you have a chance to test the alpha!
> 
> `3.7.0-alpha.3` resolved the issue for me! Was stuck on v3.2.9 only because the later versions had this issue. Thanks for the fix, hope it is merged in the stable release soon.

Well, the warning and error logged is fixed. But the problem is that the application doesn't subscribe at all. <br />
In my case particularly, I make a subscription on Page1, then on going to Page2 the previous subscription is 'completed' and second subscription request is initiated. Earlier the error `Observable cancelled prematurely` used to occur on going to Page2. That error doesn't show up now but the second subscription is not initiated at all. Can't see that in the network tab. <br />
![image](https://user-images.githubusercontent.com/48870803/177085442-67937874-370c-44ab-ae13-e13273fe8a9c.png)

**Comment by myth535** [comment]
@benjamn Thanks for your work, but the problem is still not solved, and it's worse in the new versions, because before you could at least see the error and know that something was wrong, but now not even that. Could you re-add the error message? I have a temporary fix using the error
